### ISSUE TITLE
Android APK are being zipped

### ISSUE THREAD

**Issue by ricardocostaza** [itchio/butler#232]
Hello there,

Android APK are being zipped when sent using butler. This is very bad as phone users usually can't unzip things.
The ideal behavior is to download APKs directly.
This bug is a regression from issue #199.

All the best,

**Comment by SergioAlonsoJr** [comment]
I had this problem, but it seems to have fixed itself after following these steps.

1) Deleted the Android APK channel
2) Pushed it again, but removed "quotes" from the apk directory.

Not sure what happened exactly, but I can now download the apk  and not the zip

**Comment by ricardocostaza** [comment]
> I had this problem, but it seems to have fixed itself after following these steps.
> 
> 1. Deleted the Android APK channel
> 2. Pushed it again, but removed "quotes" from the apk directory.
> 
> Not sure what happened exactly, but I can now download the apk and not the zip

Hey Sergio. Thanks for the suggestion. Unfortunately it didn't do the trick for me. Do you mind sharing the whole command line you're using? Maybe there's other small detail that plays a role on it. Thanks.

**Comment by SergioAlonsoJr** [comment]
It was something like this

`butler push path user/project:channel --if-changed`

I tried this same thing with other file (a .pdf walkthrough), and it became .zip 

I don't want to stress over this thing at this time.

**Comment by Ryhon0** [comment]
Still an issue in 2024!

**Comment by DemonikGH** [comment]
Still an issue! Can't push apk files with butler. I tried putting single apk in a folder and pushing it but it just resulted in zip with apk inside which is even worse. Version v15.26.1, built on Feb 26 2026 @ 23:39:51, ref 8557c8e5baf2db331172bd561fbba4f301af1f56. Same issue with head, built on Apr  3 2026 @ 02:03:34, ref a9ce6e298c45450c1b2ca34f5e6bf9cc64c7d35c.

**Comment by leafo** [comment]
Unfortunately butler's patching system really ins't compatible with apk files. Single files are transparently treated as a folder with a single file in it, and the output of butler is to create a zip with the pushed "folder".

Since apk files are essentially zip files, butler's patch generation isn't going to work well with subsequent zip file passed.

Doing a plain upload via the website is recommended for now until butler has a way to just do plain upload and skip patching infrastructure

https://itch.io/docs/butler/single-files.html

**Comment by DemonikGH** [comment]
Plain upload only supports 1GB files. Why is this documented then if it's not supported? https://itch.io/docs/butler/single-files.html

Also this bug has been fixed before, why did this regress? Why not just add exception for apk files by extension or by file structure? Isn't this essentially what was done before to fix this issue?

**Comment by leafo** [comment]
> Also this bug has been fixed before, why did this regress? 

I don't believe this was ever supported, butler fundamentally works a folder for patching and compression. That's what the doc you linked tries to communicate.

**Comment by DemonikGH** [comment]
Then what about discussions in #199 ? Also last part specifically talks about ability to upload individual single files.

```
If you've read all this, but you have your reasons, you can still pass a single file to butler push and it'll work transparently.

butler will behave as if you had created a folder and put your single file in it. The upload will work as usual for users of [the itch app](https://itch.io/app) (although diffs might be unusually large, if your single file is compressed or just has high entropy), and users who download from [the itch.io website](https://itch.io/) will download the single file directly (it won't be packaged in a .zip for them).
```

**Comment by leafo** [comment]
Ah, my mistake. I see now we did have separate task on the backend to extract the zip file and upload another artifact to the build with the unpacked file for direct download. In order to investigate I'll need to have the build id (or link to game page) that is experiencing the issue so I can check the build logs

**Comment by DemonikGH** [comment]
For example build 1616182. But any APK upload seems to work that way (on a separate note new --hidden flag doesn't seem to actually work and new channel wasn't hidden at least on windows bleeding edge version).

**Comment by leafo** [comment]
@DemonikGH I see an "unpacked" direct apk download file for that build. Did you try downloading it off the page before you hid it? 

I see that on the dashboard file list it currently shows the "archive" file name, but on the game's actual page it will display the available unpacked file.

It's possible the unpacked file could have still been processing (although it should go pretty fast), so in that case it would show the zip file in both places

**Comment by DemonikGH** [comment]
Yes this seems to have worked, although in the edit page it still shows it as zip archive. It did show as a zip file on download page originally.

**Comment by leafo** [comment]
there's a patch going out that will display the "unpacked" file name on the dashbaord instead of the archive one, which should make it less confusing about what file is being served. I was not able to find any bug that prevented the unpacked archive from being served on web.

**Comment by DemonikGH** [comment]
Ok there is one channel that is still stuck in zipped form. Build 1618474 didn't change to apk. Reuploading same file with the same command to the same channel forced it to unpack (build 1619750).

**Comment by DemonikGH** [comment]
Now the same happened with build 1618485. It was apk but now it changed to zip.

**Comment by leafo** [comment]
Thanks for the build IDs @DemonikGH , I believe I found a bug in our http library that was preventing the backend task to upload the unpacked archive from completing (so only the full zip file was available), testing pushing out some fixes today. I'll see about re-queuing those jobs

**Comment by leafo** [comment]
All outstanding unpacked jobs have been reprocessed. Closing this out as the issue should be fixed now. Thanks for the reports everyone.
